### ISSUE TITLE
Fix false positives for :is() and :where() in *-specificity

### ISSUE THREAD

**Issue by jonathantneal** [stylelint/stylelint#5065]
I noticed a false positive error coming from a pending release of normalize that uses `:where()`.

Here is a sample of the CSS being flagged.

```css
/**
 * Remove the margin on nested lists in Chrome, Edge, and Safari.
 */

:where(dl :is(dl, ol, ul), :is(ol, ul) dl) {
  margin-block-end: 0;
  margin-block-start: 0;
}

/**
 * 1. Add the correct box sizing in Firefox.
 * 2. Correct the inheritance of border color in Firefox.
 */

:where(hr) {
  box-sizing: content-box; /* 1 */
  color: inherit; /* 2 */
  height: 0; /* 1 */
}

/**
 * 1. Correct the inheritance and scaling of font size in all browsers.
 * 2. Correct the odd `em` font sizing in all browsers.
 */

:where(pre) {
  font-family: monospace, monospace; /* 1 */
  font-size: 1em; /* 2 */
}
```

And here are the errors reported from that sample:

```
✖  Expected selector ":where(hr)" to come     no-descending-specificity
   before selector ":where(dl :is(dl, ol,
   ul), :is(ol, ul) dl)"
✖  Expected selector ":where(pre)" to come    no-descending-specificity
   before selector ":where(dl :is(dl, ol,
   ul), :is(ol, ul) dl)"
✖  Expected selector ":where(b, strong)" to   no-descending-specificity
   come before selector ":where(dl :is(dl,
   ol, ul), :is(ol, ul) dl)"
```

The `:where()` CSS pseudo-class function takes a selector list as its argument, and selects any element that can be selected by one of the selectors in that list.

This is useful for writing large selectors in a more compact form, and also for writing selectors with 0 specificity.

> Which rule, if any, is the bug related to?

```
:where()
```

> What code is needed to reproduce the bug?

☝️See the top comment.

> What stylelint configuration is needed to reproduce the bug?

e.g.

```json
{
  "extends": "stylelint-config-standard"
}
```

> Which version of stylelint are you using?

e.g. `13.8.0`

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

```
stylelint evergreen.css
```

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

No, `:switch()` ships Firefox 78, Safari 14, & Chrome 88 📌.

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors did you get)?

☝️See the top comment.

> Additional notes:

📌 _Chrome is v87 at the time of this writing. Firefox shipped v78 in June, and Safari shipped v14 in September_

**Comment by jeddy3** [comment]
@jonathantneal Thanks for the clear write-up and for using the template.

I'll label as upstream as the `no-descending-specificity` rule relies on the [`specificity` npm package](https://github.com/keeganstreet/specificity) under the hood.

You may want to turn off the rule in your config object until someone finds time to update that package.

---

[MDN ref](https://developer.mozilla.org/en-US/docs/Web/CSS/:where) & [spec ref](https://drafts.csswg.org/selectors-4/#zero-matches) for `:where()`.

**Comment by malsf21** [comment]
Just poking around - looks like the specificity npm package hasn't been updated in over two years. Should we submit an issue and/or fork a copy to the styelint org? More than happy to do some writeups if needed!

**Comment by jeddy3** [comment]
Let's submit an issue upstream.

**Comment by mattxwang** [comment]
As an update, it seems like `specificity` doesn't support Selectors Level 4 syntax like the `:where` pseudoclass, and implementing it would require a rewrite of the package as-is (it currently uses regex, where maybe we might need to use PostCSS and an AST for the newer, complicated selectors).

Any thoughts on how to proceed? I am not particularly familiar with implementing a solution to this problem, but more than happy to give it a shot!

**Comment by jeddy3** [comment]
Thanks for digging into this some more and asking upstream. It does sound like the long term solution is an AST-based rewrite of a specificity calculator.

In the [`selector-max-specificity` rule](https://github.com/stylelint/stylelint/tree/master/lib/rules/selector-max-specificity) we do some stuff for `:not` and and `:matches` (the old name for `:is`):

https://github.com/stylelint/stylelint/blob/4f4bdf94edba76e65e9ff86c862e6435f321e0a7/lib/rules/selector-max-specificity/index.js#L90-L102

Perhaps a temporary solution is repurposing some of that approach in the `no-descending-specificity` rule, i.e. combining the selector parser with the `specificity` library.  It may get messy, though, but it's probably worth an initial investigation.

**Comment by jeddy3** [comment]
> Perhaps this could be considered.?

@bramus Definitely. 

It's great to see the bundle size decrease in https://github.com/bramus/js-specificity/pull/4. It's still quite large, though. The [bulk of the package](https://unpkg.com/@bramus/specificity@1.0.4/dist/index.cjs.js) seems to be syntax for values and media features. As you're [only using the parser in a `SelectorList` context](https://github.com/bramus/js-specificity/blob/main/src/index.js#L139-L141), do you know if it's possible to import only that part of the parser?

If that's not possible, we can either:

- use the package as-is, which is 50k minified
- port the [specificity logic into our own util](https://github.com/bramus/js-specificity/blob/main/src/index.js) and adapt it to use the [PostCSS Selector Parser](https://www.npmjs.com/package/postcss-selector-parser), which we already bundle in Stylelint

For the latter, we did [something similar for the `named-grid-areas-no-invalid` rule](https://github.com/stylelint/stylelint/issues/5155) i.e. took an existing package and ported it to Stylelint to use the PostCSS Value Parser.

**Comment by lahmatiy** [comment]
> The [bulk of the package](https://unpkg.com/@bramus/specificity@1.0.4/dist/index.cjs.js) seems to be syntax for values and media features.

@jeddy3 You examined js-specificity 1.0.4, a version before the reducing of bundle size. The version with smaller bundle size is [1.0.5](https://unpkg.com/@bramus/specificity@1.0.5/dist/index.cjs.js). But, yeah, this version still contains a general purpose CSS parser. According to my quick and dirty experiment, making the parser to be specialized on a selectors parsing only can cut bundle size by ~50%.

**Comment by jeddy3** [comment]
> @jeddy3 You examined js-specificity 1.0.4, a version before the reducing of bundle size. 

Sorry, my mistake.

> According to my quick and dirty experiment, making the parser to be specialized on a selectors parsing only can cut bundle size by ~50%

This would be great!

@bramus Thanks for posing the question upstream in https://github.com/csstree/csstree/issues/183

With [context-specialised parsers in the pipeline](https://github.com/csstree/csstree/issues/183#issuecomment-1050372897), I think we can update the `no-descending-specificity` and `selector-max-specificity` rules to use your package.

It'll fix our bugs with `:is()` and `:where()`; two methods of managing specificity that are becoming more common.

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to fix a bug in a rule](https://stylelint.io/developer-guide/rules#fix-a-bug-in-a-rule) in the Developer guide.

**Comment by bramus** [comment]
@jeddy3 FYI: I'm currently working on v2 of `@bramus/js-specificity`.

The main changes are that it plays nice with SelectorLists (i.e. it'll always return an **array** of specificities, similar to how the `specificity` package does it).

It also comes with some helper functions to filter/sort/compare specificities. Migrating to it should be fairly easy that way.

You can check a preview of it on the `dev` branch at https://github.com/bramus/js-specificity/tree/dev _(the header image isn't up-to-date, but the examples in the README are)_

I'm open to feedback on the returned format, in order to make your _(possible)_ migration a breeze. Feel free to leave a reply on https://github.com/bramus/js-specificity/issues/2 in that case.

**Comment by ybiquitous** [comment]
I've opened PR #6083 using `@bramus/specificity`.

**Comment by ybiquitous** [comment]
@romainmenke Thanks, the [`@csstools/selector-specificity`](https://www.npmjs.com/package/@csstools/selector-specificity) package as you suggest looks great! It may prevent Stylelint from increasing its package size.

I'll open a new PR using it.
